xde/src/route.rs

```diff
@@ -0,0 +1,587 @@
+            Some(route) if route.is_valid(t) => return route.into_route(xde),
+            _ => {}
+        }
+
+        // Cache miss: intent is to now ask illumos, then insert.
+        let mut route_cache = self.0.write();
+        let space_remaining = route_cache.len() < MAX_CACHE_ENTRIES;
+
+        // Someone else may have written while we were taking the lock.
+        // DO NOT waste time if there's a good route.
+        let maybe_route = route_cache.entry(key);
+        let entry_exists = match &maybe_route {
+            Entry::Occupied(e) if e.get().is_valid(t) => {
+                return e.get().into_route(xde);
+            }
+            Entry::Occupied(_) => true,
+            _ => false,
+        };
+
+        // We've had a definitive flow miss, but we need to cap the cache
+        // size to prevent excessive modification latencies at high flow
+        // counts.
+        // If full and we have no old entry to update, drop the lock and do
+        // not insert.
+        // XXX: Want to profile in future to see if LRU expiry is
+        //      affordable/sane here.
+        // XXX: A HashMap would exchange insert cost for lookup.
+        if entry_exists || space_remaining {
+            // `next_hop` might fail for myriad reasons, but we still
+            // send the packet on an underlay device depending on our
+            // progress. However, we do not want to cache bad mappings.
+            match (maybe_route, next_hop(&key, xde)) {
+                (Entry::Vacant(slot), Ok(route)) => {
+                    slot.insert(route.cached(xde, t));
+                    route
+                }
+                (Entry::Occupied(mut slot), Ok(route)) => {
+                    slot.insert(route.cached(xde, t));
+                    route
+                }
+                (_, Err(dev)) => Route::zero_addr(dev),
+            }
+        } else {
+            drop(route_cache);
+            match next_hop(&key, xde) {
+                Ok(route) => route,
+                Err(dev) => Route::zero_addr(dev),
+            }
+        }
+    }
+
+    /// Discards any cached route entries which have been present
+    /// for longer than `REMOVE_ROUTE_LIFETIME`.
+    pub fn remove_routes(&self) {
+        let mut route_cache = self.0.write();
+
+        let t = Moment::now();
+        route_cache.retain(|_, v| v.is_retained(t));
+    }
+}
+
+/// An underlay routing destination and flow-dependent entropy.
+#[derive(Copy, Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
+pub struct RouteKey {
+    pub dst: Ipv6Addr,
+    pub l4_hash: Option<u32>,
+}
+
+/// Cached representation of [`Route`].
+#[derive(Copy, Clone, Debug)]
+pub struct CachedRoute {
+    pub src: EtherAddr,
+    pub dst: EtherAddr,
+    pub underlay_idx: u8,
+    pub timestamp: Moment,
+}
+
+impl CachedRoute {
+    fn is_valid(&self, t: Moment) -> bool {
+        t.delta_as_millis(self.timestamp)
+            <= EXPIRE_ROUTE_LIFETIME.as_millis() as u64
```

Comment:
Should we detect rollover here with `try_into` or similar and return false if we detect rollover? Same for `is_retained`. Or if this is a smaller type, using `into` would make it more obvious that this is truncation/rollover safe.

Reply:
Done; std/core represents millisecond counds as `u128` while our `hrtime_t` is `i64` (nanoseconds). There isn't a  practical issue here given rollover times, but I've put `u128::from` on the LHS to make it obvious we're not doing any truncation.